sema/discovery/url_to_graph.py

```diff
@@ -0,0 +1,152 @@
+
+log = logging.getLogger(__name__)
+
+RDF_MIME_TO_FORMAT = {
+    "application/ld+json": "json-ld",
+    "text/turtle": "turtle",
+}
+
+
+def ctype_to_rdf_format(ctype: str) -> str:
+    return RDF_MIME_TO_FORMAT.get(ctype, None)
+
+
+def get_graph_for_format(subject_url: str, formats: str, graph: Graph = None):
+    """
+    Discover triples describing the subject (assumed at subject_url)
+    and add them to the graph
+
+    :param subject_url: url (originally assumed from <uri>)
+    pointing to the subject to be discovered
+    :type subject_url: str
+    :param g: graph to be filled
+    :type g: rdflib.Graph
+    :param format: indicating what format should be retrieved json-ld, turtle
+    :type form: str
+    :returns: the graph whith added discovered triples
+    :rtype: rdflib.Graph
+    """
+
+    if subject_url is None:
+        return None
+
+    if graph is None:
+        graph = Graph()  # create a fresh graph if you don't have it yet
+
+    total_retry = 8
+    session = requests.Session()
+    retry = Retry(
+        total=total_retry,
+        backoff_factor=0.4,
+        status_forcelist=[500, 502, 503, 504, 429],
+    )
+    adapter = HTTPAdapter(max_retries=retry)
+    session.mount("http://", adapter)
+    session.mount("https://", adapter)
+
+    triples_found = False
+
+    ACCEPTABLE_MIMETYPES = {
+        "application/ld+json",
+        "text/turtle",
+        "application/json",
+    }
+
+    for format in formats:
+        headers = {"Accept": format}
+        log.debug(f"requesting {subject_url} with {headers=}")
+        r = session.get(subject_url, headers=headers)
+        mime_type, options = cgi.parse_header(r.headers["Content-Type"])
+        log.debug(f"got {r.status_code=} {mime_type=}")
+
+        if r.status_code == 200 and bool(mime_type in ACCEPTABLE_MIMETYPES):
+            triples_found = True
+            try:
+                # if mimetype is application/json assume application/ld+json
+                # to satisfy the known formats of rdflib.parser
+                if mime_type == "application/json":
+                    mime_type = "application/ld+json"
+                if mime_type == "application/octet-stream":
+                    mime_type = "text/turtle"
+                graph.parse(
+                    data=r.text, format=mime_type, publicID=subject_url
+                )
+
+            except Exception as e:
+                log.warning(
+                    f"failed to parse {subject_url} in {format=} error: {e}"
+                )
+
+            finally:
+                return graph
+
+    if not triples_found:
+        # perform a check in the html to
+        # see if there is any link to fair signposting
+        # perform request to uri with accept header text/html
+        headers = {"Accept": "text/html"}
+        r = session.get(subject_url, headers=headers)
+        if r.status_code == 200 and "text/html" in r.headers["Content-Type"]:
```

Comment:
make an issue to enhance this kind of tests -- I rather see us using a lib-function like [`cgi.parse_header()`](https://docs.python.org/3/library/cgi.html) for proper http header parser - allowing actual match to the mimetype, and not an unintendent (but unlikely) stray match

Reply:
Issue taken up in #65